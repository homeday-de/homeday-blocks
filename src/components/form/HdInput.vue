<template>
  <div
    :class="fieldClasses"
    class="field field--input"
  >
    <img
      v-if="icon"
      :src="icon"
      role="presentation"
      class="field__icon"
    >
    <input
      :autocomplete="autocomplete"
      :value="value"
      :type="currentType"
      :id="name"
      :name="name"
      :placeholder="isActive && placeholder !== undefined ? placeholder : ''"
      :required="required"
      :autofocus="autofocus"
      :disabled="disabled"
      class="field__input"
      ref="input"
      data-lpignore="true"
      @input="handleInput"
      @focus="handleFocus"
      @blur="handleBlur"
    />
    <label
      v-if="label"
      :for="name"
      class="field__label"
    >
      {{ label }}
    </label>
    <p
      v-if="error"
      class="field__error"
    >
      {{ error }}
    </p>
    <p
      v-else-if="helper"
      class="field__error field__error--helper"
      v-html="helper"
    />
    <span class="field__border"/>
  </div>
</template>

<script>
import merge from 'lodash/merge';
import { getMessages } from 'homeday-blocks/src/lang';
import {
  email as validateEmail,
  date as validateDate,
} from 'homeday-blocks/src/services/formValidation';

export default {
  name: 'HdInput',
  props: {
    label: {
      type: String,
      default: '',
    },
    name: {
      type: String,
      required: true,
    },
    type: {
      type: String,
      default: 'text',
    },
    value: {
      type: [String, Number],
      default: '',
    },
    placeholder: {
      type: String,
      default: '',
    },
    required: {
      type: Boolean,
      default: false,
    },
    autocomplete: {
      type: String,
      default: 'on',
    },
    autofocus: {
      type: Boolean,
      default: false,
    },
    min: {
      type: Number,
      default: undefined,
    },
    max: {
      type: Number,
      default: undefined,
    },
    lang: {
      type: String,
      default: 'de',
    },
    texts: {
      type: Object,
      default: () => ({}),
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    icon: {
      type: String,
      default: '',
    },
    customRules: {
      type: Array,
      default: () => [],
      validator: rulesProvided => rulesProvided.every(
        ({ validate, errorMessage }) => typeof validate === 'function' && typeof errorMessage === 'string',
      ),
    },
  },
  data() {
    return {
      currentType: this.type || 'text',
      isActive: undefined,
      isValid: undefined,
      error: null,
      helper: null,
    };
  },
  computed: {
    t() {
      return merge(getMessages(this.lang), this.texts);
    },
    isEmpty() {
      return this.value === null || this.value === undefined || this.value === '';
    },
    fieldClasses() {
      return {
        'field--active': this.isActive,
        'field--filled': !this.isEmpty,
        'field--invalid': this.isValid === false,
        'field--disabled': this.disabled,
        'field--hasIcon': this.icon,
      };
    },
  },
  watch: {
    value() {
      this.validate();
    },
    type(type) {
      this.currentType = type;
    },
  },
  methods: {
    clearInput() {
      this.$refs.input.focus();
      this.hideError();
      this.$emit('input', '');
    },
    handleFocus() {
      this.isActive = true;
      this.$emit('focus');
    },
    handleBlur() {
      this.isActive = false;
      this.validate();
      this.$emit('blur');
    },
    handleInput(e) {
      let newValue = e.target.value;

      if (this.currentType === 'number' && newValue !== '') {
        newValue = parseFloat(newValue);

        if (typeof this.min === 'number' && newValue < this.min) {
          newValue = this.min;
        } else if (typeof this.max === 'number' && newValue > this.max) {
          newValue = this.max;
        }

        if (newValue === this.value) {
          this.$nextTick(this.$forceUpdate);
        }
      }

      this.$emit('input', newValue);
    },
    showError(errorMessage) {
      this.error = errorMessage;
      this.isValid = false;
    },
    showHelper(message) {
      this.helper = message;
    },
    hideError() {
      this.isValid = true;
      this.error = null;
    },
    validate() {
      if (this.required && this.isEmpty) {
        this.showError(this.t.FORM.VALIDATION.REQUIRED);
      } else if (this.currentType === 'email' && !validateEmail(this.value)) {
        this.showError(this.t.FORM.VALIDATION.INVALID_EMAIL);
      } else if (this.currentType === 'date' && !validateDate(this.value)) {
        this.showError(this.t.FORM.VALIDATION.INVALID_DATE);
      } else if (this.customRules.length && !this.isEmpty) {
        this.validateCustomRules();
      } else {
        this.hideError();
      }

      return !this.error;
    },
    validateCustomRules() {
      const firstFailingRule = this.customRules.find(({ validate }) => !validate(this.value));
      if (firstFailingRule) {
        this.showError(firstFailingRule.errorMessage);
      } else {
        this.hideError();
      }
    },
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="scss">
@import 'homeday-blocks/src/styles/mixins.scss';
@import 'homeday-blocks/src/styles/inputs.scss';

.field {
  &__error {
    width: 100%;
    text-align: left;
    &--helper {
      display: block;
      color: getShade($quaternary-color, 80);
    }
  }
}
</style>
