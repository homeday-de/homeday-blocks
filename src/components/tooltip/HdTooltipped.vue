<template>
  <div
    v-if="component !== undefined"
    v-hd-tooltip="{ value, direction }"
    :is="component"
    v-bind="props"
  ></div>
  <span v-else>
    <slot></slot>
  </span>
</template>

<script>
export default {
  name: 'hd-tooltipped',
  props: {
    value: String,
    direction: String,
    component: Object,
    props: Object,
  },
  computed: {},
};
</script>
