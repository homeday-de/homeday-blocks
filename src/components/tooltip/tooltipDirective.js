/* eslint-disable no-use-before-define, prefer-destructuring */

function getFormattedBinding(binding) {
  // in case of: v-tooltip.left="'test'"
  let { value, modifiers } = binding;
  // in case of: v-tooltip="{value: 'test', direction: 'left'}"
  if (typeof binding.value === 'object') {
    value = binding.value.value;
    modifiers = {
      [binding.value.direction]: true,
    };
  }
  return { value, modifiers };
}
export default {
  componentUpdated(el, binding) {
    if (binding.value !== binding.oldValue) {
      el.setAttribute('aria-label', getFormattedBinding(binding).value);
    }
  },
  bind(el, binding, vnode) {
    const tooltipElm = window.tooltipElement;
    if (tooltipElm === undefined) return;
    let tooltipRect = tooltipElm.getBoundingClientRect();

    const eventsIn = ['mouseenter', 'focus'];
    const eventsOut = ['mouseleave', 'blur'];

    const { value, modifiers } = getFormattedBinding(binding);

    const parent = document.body;
    const SPACING_HORIZONTAL = 8;
    const SPACING_VERTICAL = 2;

    eventsIn.forEach((event) => el.addEventListener(event, show));
    eventsOut.forEach((event) => el.addEventListener(event, hide));

    parent.style.position = parent.style.position || 'relative'; // To position the tooltip reltively to its parent
    parent.appendChild(tooltipElm);
    el.setAttribute('aria-label', value);

    function show() {
      tooltipElm.innerText = el.getAttribute('aria-label');
      tooltipElm.classList.add('show');
      reflow();
    }
    function hide() {
      tooltipElm.classList.remove('show');
    }
    function reflow() {
      tooltipElm.style.top = 'auto';
      tooltipElm.style.left = 'auto';
      tooltipElm.style.right = 'auto';
      vnode.context.$nextTick(() => {
        const elRect = el.getBoundingClientRect();
        const parentRect = parent.getBoundingClientRect();
        refreshTooltipRect();

        if (modifiers.left) {
          tooltipElm.classList.add('left');
          tooltipElm.style.right = `${(parentRect.right - elRect.right) + elRect.width + SPACING_HORIZONTAL}px`;
          refreshTooltipRect();
          tooltipElm.style.top = `${(elRect.top - parentRect.top - (tooltipRect.height / 2)) + (elRect.height / 2)}px`;
        } else {
          tooltipElm.classList.remove('left');
          tooltipElm.style.top = `${(elRect.top - parentRect.top) + elRect.height + SPACING_VERTICAL}px`;
          const left = (elRect.left - parentRect.left - (tooltipRect.width / 2)) + (elRect.width / 2);
          tooltipElm.style.left = `${left}px`;
          // To void overflow
          if (left < 0) {
            tooltipElm.style.left = '0px';
          }
        }
      });
    }
    function refreshTooltipRect() {
      tooltipRect = tooltipElm.getBoundingClientRect();
    }
  },
};
