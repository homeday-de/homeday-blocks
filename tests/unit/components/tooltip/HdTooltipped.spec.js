import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import Vue from 'vue';
import HdTooltipped from '@/components/tooltip/HdTooltipped.vue';
import HdInput from '@/components/form/HdInput.vue';
import tooltipDirective from '@/components/tooltip/tooltipDirective';

// required to properly solve HDTooltipped
Vue.directive('hd-tooltip', tooltipDirective);

const wrapperBuilder = wrapperFactoryBuilder(HdTooltipped);

describe('HdTooltipped', () => {
  it('renders empty', () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders full', () => {
    const wrapper = wrapperBuilder({
      props: {
        component: HdInput,
        value: 'Lorem ipsum',
        direction: 'left',
        props: {
          label: 'dolor sit ammet',
          type: 'text',
          name: 'testInput',
        },
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
