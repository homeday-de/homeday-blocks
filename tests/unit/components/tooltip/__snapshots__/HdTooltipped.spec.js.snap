// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HDTooltipped the component is rendered 1`] = `<span></span>`;

exports[`HDTooltipped the component is rendered 2`] = `
<div class="field field--input">
  <!----> <input autocomplete="on" type="text" id="testInput" name="testInput" placeholder="" data-lpignore="true" class="field__input"> <label for="testInput" class="field__label">
    dolor sit ammet
  </label>
  <!----> <span class="field__border"></span></div>
`;
